example/common/flat_stream.hpp

```diff
@@ -0,0 +1,567 @@
+std::size_t
+flat_stream<NextLayer>::
+read_some(MutableBufferSequence const& buffers)
+{
+    static_assert(boost::beast::is_sync_read_stream<next_layer_type>::value,
+        "SyncReadStream requirements not met");
+    static_assert(boost::asio::is_mutable_buffer_sequence<
+        MutableBufferSequence>::value,
+            "MutableBufferSequence requirements not met");
+    boost::system::error_code ec;
+    auto n = read_some(buffers, ec);
+    if(ec)
+        BOOST_THROW_EXCEPTION(boost::system::system_error{ec});
+    return n;
+}
+
+template<class NextLayer>
+template<class MutableBufferSequence>
+std::size_t
+flat_stream<NextLayer>::
+read_some(MutableBufferSequence const& buffers, boost::system::error_code& ec)
+{
+    return stream_.read_some(buffers, ec);
+}
+
+template<class NextLayer>
+template<class ConstBufferSequence>
+std::size_t
+flat_stream<NextLayer>::
+write_some(ConstBufferSequence const& buffers)
+{
+    static_assert(boost::beast::is_sync_write_stream<next_layer_type>::value,
+        "SyncWriteStream requirements not met");
+    static_assert(boost::asio::is_const_buffer_sequence<
+        ConstBufferSequence>::value,
+            "ConstBufferSequence requirements not met");
+    boost::system::error_code ec;
+    auto n = write_some(buffers, ec);
+    if(ec)
+        BOOST_THROW_EXCEPTION(boost::system::system_error{ec});
+    return n;
+}
+
+template<class NextLayer>
+template<class ConstBufferSequence>
+std::size_t
+flat_stream<NextLayer>::
+write_some(
+    ConstBufferSequence const& buffers,
+    boost::system::error_code& ec)
+{
+    auto const first = boost::asio::buffer_sequence_begin(buffers);
+    auto const last = boost::asio::buffer_sequence_end(buffers);
+    auto const result = coalesce(first, last, coalesce_limit);
+    auto const b = buffers_iterator_range(first, result.last);
+    if(result.copy)
+    {
+        std::unique_ptr<char[]> p{new char[result.size]};
+        boost::asio::buffer_copy(
+            boost::asio::buffer(p.get(), result.size), b);
+        return stream_.write_some(
+            boost::asio::buffer(p.get(), result.size), ec);
+    }
+
+    return stream_.write_some(b, ec);
+}
+
+template<class NextLayer>
+template<
+    class MutableBufferSequence,
+    class ReadHandler>
+BOOST_ASIO_INITFN_RESULT_TYPE(
+    ReadHandler, void(boost::system::error_code, std::size_t))
+flat_stream<NextLayer>::
+async_read_some(
+    MutableBufferSequence const& buffers,
+    ReadHandler&& handler)
+{
+    static_assert(boost::beast::is_async_read_stream<next_layer_type>::value,
+        "AsyncReadStream requirements not met");
+    static_assert(boost::asio::is_mutable_buffer_sequence<
+            MutableBufferSequence >::value,
+        "MutableBufferSequence  requirements not met");
+    return stream_.async_read_some(
+        buffers, std::forward<ReadHandler>(handler));
+}
+
+template<class NextLayer>
+template<
+    class ConstBufferSequence,
+    class WriteHandler>
```

Comment:
`WriteHandler` is a CompletionToken, rather than a CompletionHandler, so the template parameter name may be misleading.

Reply:
Yeah, that's a fair point. This could go both ways. Using `WriteHandler` instead of `CompletionToken` gives the reader a tiny bit more information. And it also allows the documentation toolchain to hyperlink the template parameter name, see:
https://www.boost.org/doc/libs/1_67_0/libs/beast/doc/html/beast/ref/boost__beast__http__async_write_some.html

Beast assumes that the user already knows Asio. So `WriteHandler` is more informative.